oidc_client/utils/utils.py

```diff
@@ -0,0 +1,133 @@
+
+    return response
+
+
+async def request_token(code):
+    """Request token from AAI."""
+    LOG.debug('Requesting token.')
+
+    auth = aiohttp.BasicAuth(login=CONFIG.aai['client_id'], password=CONFIG.aai['client_secret'])
+    data = {
+        'grant_type': 'authorization_code',
+        'code': code,
+        'redirect_uri': CONFIG.aai['url_callback']
+    }
+
+    # Set up client authentication for request
+    async with aiohttp.ClientSession(auth=auth) as session:
+        # Send request to AAI
+        async with session.post(CONFIG.aai['url_token'], data=data) as response:
+            LOG.debug(f'AAI response status: {response.status}.')
+            # Validate response from AAI
+            if response.status == 200:
+                # Parse response
+                result = await response.json()
+                # Look for access token
+                if 'access_token' in result:
+                    LOG.debug('Access token received.')
+                    return result['access_token']
+                else:
+                    LOG.error('AAI response did not contain an access token.')
+                    raise web.HTTPBadRequest(text='AAI response did not contain an access token.')
+            else:
+                LOG.error(f'Token request to AAI failed: {response}.')
+                LOG.error(await response.json())
+                raise web.HTTPBadRequest(text=f'Token request to AAI failed: {response.status}.')
+
+
+async def query_params(request):
+    """Parse query string params from path."""
+    LOG.debug('Parse query params from AAI response.')
+
+    # Response from AAI must have the query params `state` and `code`
+    if 'state' in request.query and 'code' in request.query:
+        LOG.debug('AAI response contained the correct params.')
+        return {'state': request.query['state'], 'code': request.query['code']}
+    else:
+        LOG.error(f'AAI response is missing mandatory params, received: {request.query}')
+        raise web.HTTPBadRequest(text='AAI response is missing mandatory parameters.')
+
+
+async def check_bona_fide(token):
+    """Check if user is recognised as a Bona Fide researcher."""
+    LOG.debug('Checking Bona Fide status.')
+
+    terms = False
+    status = False
+    headers = {"Authorization": f"Bearer {token}"}
+
+    async with aiohttp.ClientSession() as session:
+        # Send request to AAI
+        async with session.get(CONFIG.aai['url_userinfo'], headers=headers) as response:
+            LOG.debug(f'AAI response status: {response.status}.')
+            # Validate response from AAI
+            if response.status == 200:
+                # Parse response
+                result = await response.json()
+                # Check for Bona Fide values
+                ga4gh = result.get('ga4gh', {})
+                if 'AcceptedTermsAndPolicies' in ga4gh:
+                    for accepted_terms in ga4gh["AcceptedTermsAndPolicies"]:
```

Comment:
We are looking for at least one of  `AcceptedTermsAndPolicies` to be equal to `CONFIG.elixir['bona_fide_value']` ?
The same goes for `ResearcherStatus` ?
Can there be multiple and if one of them is not equal we ignore it ?

Reply:
Maybe somewhere a link to actual specification: https://github.com/ga4gh-duri/ga4gh-duri.github.io/blob/master/researcher_ids/RI_Claims_V1.md

Reply:
[AcceptedTermsAndPolicies](https://github.com/ga4gh-duri/ga4gh-duri.github.io/blob/master/researcher_ids/RI_Claims_V1.md#ga4ghacceptedtermsandpolicies) must contain `https://doi.org/10.1038/s41431-018-0219-y` to signify that the user has agreed to the Bona Fide contract.

[ResearcherStatus](https://github.com/ga4gh-duri/ga4gh-duri.github.io/blob/master/researcher_ids/RI_Claims_V1.md#ga4ghresearcherstatus) must contain `https://doi.org/10.1038/s41431-018-0219-y` to signify that the user has been vouched for by another esteemed researcher for upholding good ethics and intents.

For gaining Bona Fide attribute to services, both of these requirements must be met. It doesn't matter what other objects they may contain.

Reply:
I was thinking of the use case where we need to check multiple of them at the same time, it changes a lot the logic of that function.

Reply:
There are two options here:
* leave it as is with the mention in the docs that this has been designed for checking only Bona Fide for now and do an issue on it to modify it
* make the change now

i assume the former would be more suitable.

Reply:
The values in the structure we need to check for are
```
{
  "ga4gh": {
    "AcceptedTermsAndPolicies": [
        {
          "value": "string"
        }
    ],
    "ResearcherStatus": [
        {
          "value": "string"
        }
    ]
  }
}
```
So it's `obj obj list obj str` `{{[{""}]}}` and for loop seemed like a good way to go through those. Would it be possible to check it via some kind of schema like in https://github.com/CSCfi/beacon-python/blob/master/beacon_api/utils/validate.py#L67 ?

Reply:
sure, but preferably in a different issue.

Reply:
it seems: https://github.com/ga4gh-duri/ga4gh-duri.github.io/blob/master/researcher_ids/RI_Claims_V1.md#registered-access that checking those values for registered access is enough. We can have an issue to make this implementation more flexible later down the road, for now this is enough.